torchtitan/models/deepseek_v3/model/state_dict_adapter.py

```diff
@@ -52,6 +60,278 @@ def __init__(self, model_args: DeepSeekV3ModelArgs, hf_assets_path: str | None):
+
+        # Step 4: Create individual expert DTensors
+        assert isinstance(
+            grouped_expert_weight, DTensor
+        ), "Expected DTensor for grouped expert weight"
+
+        local_grouped_weights = grouped_expert_weight._local_tensor
+        expected_local_experts = end_index - start_index
+
+        if local_grouped_weights.shape[0] != expected_local_experts:
+            raise ValueError(
+                f"Local tensor shape mismatch: expected {expected_local_experts} experts, "
+                f"got {local_grouped_weights.shape[0]}"
+            )
+
+        local_expert_tensors = {}
+        for expert_id in range(start_index, end_index):
+            expert_key = abstract_key.format(layer_id, expert_id)
+            local_expert_index = expert_id - start_index
+
+            # Extract individual expert weight and add batch dimension temporarily
+            expert_weight = local_grouped_weights[local_expert_index, :, :].unsqueeze(0)
+
+            # Create DTensor and remove batch dimension (experts dimension is removed)
+            expert_dtensor = DTensor.from_local(
+                expert_weight, device_mesh, new_placements, run_check=False
+            ).squeeze(0)
+
+            local_expert_tensors[expert_key] = expert_dtensor
+
+        return local_expert_tensors
+
+    def _concatenate_local_expert_weights(
+        self,
+        expert_weights_by_layer: dict[str, Any],
+        abstract_key: str,
+        device_mesh: DeviceMesh,
+    ) -> torch.Tensor:
+        """
+        Try to concatenate the weights of separate experts into GroupedExperts weights.
+        """
+        for layer in expert_weights_by_layer.keys():
+            # If we have all the experts for this abstract_key, concatenate them
+            experts = expert_weights_by_layer[layer][abstract_key]
+            expected_n_experts = (
+                self.local_experts_indices[abstract_key][1]
+                - self.local_experts_indices[abstract_key][0]
+            )
+            if len(experts) == expected_n_experts:
+                sorted_expert_ids = sorted(experts.keys())
+                sorted_experts = [experts[i] for i in sorted_expert_ids]
+                local_tensor = torch.stack(sorted_experts, dim=0)._local_tensor
+
+                assert (
+                    abstract_key in self.grouped_expert_weight_placements
+                    and abstract_key in self.grouped_expert_weight_shape
+                ), "GroupedExperts weight metadata (placements, shape) can not be None!"
+
+                stacked_dtensor = DTensor.from_local(
+                    local_tensor,
+                    device_mesh,
+                    self.grouped_expert_weight_placements[abstract_key],
+                    run_check=False,
+                )
+
+                # Remove these experts from the tracking dict to free memory
+                del expert_weights_by_layer[layer][abstract_key]
```

Comment:
Do we actually free the memory? Isn't that weight still tracked by `hf_state_dict`?

Reply:
No the memory is not removed. This part of memory is still tracked by other variable.